File: security-assessment-dashboard/backend/schemas/execution.py
"""Request/response schemas for the Assessment Execution Engine."""

import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from backend.models.enums import ToolExecutionStatus


class ToolExecutionOptionsSchema(BaseModel):
    """Per-tool planning options for one ``/execute`` call: a Scan Profile id and its overrides.

    Only meaningful for a plugin with a Scan Profile system (Nmap is the
    reference implementation) — naming a profile for a tool without one is
    a request-shape error, not a silent no-op.
    """

    profile_id: str | None = None
    advanced_options: dict | None = None


class ExecuteRequest(BaseModel):
    """Request body for ``POST /assessments/{id}/execute``."""

    tool_names: list[str] = Field(min_length=1, description="Plugin/tool ids to run, e.g. ['nmap'].")
    target_ids: list[uuid.UUID] | None = Field(
        default=None, description="Subset of the assessment's enabled targets. Omit to select all enabled targets."
    )
    tool_options: dict[str, ToolExecutionOptionsSchema] | None = Field(
        default=None, description="Keyed by tool name, e.g. {'nmap': {'profile_id': 'tcp_full'}}."
    )


class JobRead(BaseModel):
    """One job -- one ``ToolExecution`` row."""

    id: uuid.UUID
    assessment_id: uuid.UUID
    target_id: uuid.UUID
    target_value: str
    tool_id: uuid.UUID
    tool_name: str
    status: ToolExecutionStatus
    status_message: str | None
    started_at: datetime | None
    completed_at: datetime | None
    duration: float | None
    return_code: int | None
    retry_count: int
    log_path: str | None
    profile_id: str | None
    generated_command: list[str] | None
    created_at: datetime


class ExecuteResponse(BaseModel):
    """Result of planning and queuing one execution run."""

    assessment_id: uuid.UUID
    jobs: list[JobRead]
    queued_count: int
    skipped_count: int


class JobLogsResponse(BaseModel):
    """One job's combined log file, optionally filtered/tailed."""

    job_id: uuid.UUID
    lines: list[str]
    log_path: str | None


class AssessmentProgress(BaseModel):
    """Live counts of an assessment's jobs by status, plus what's running right now."""

    assessment_id: uuid.UUID
    total: int
    pending: int
    queued: int
    preparing: int
    running: int
    completed: int
    failed: int
    cancelled: int
    timeout: int
    skipped: int
    percent_complete: float = Field(description="Completed+failed+cancelled+timeout+skipped, as a percent of total.")
    current_jobs: list[JobRead]

File: security-assessment-dashboard/frontend/src/types/execution.ts
export type JobStatus =
  | "pending"
  | "queued"
  | "preparing"
  | "running"
  | "completed"
  | "failed"
  | "cancelled"
  | "timeout"
  | "skipped";

export const ACTIVE_JOB_STATUSES: JobStatus[] = ["pending", "queued", "preparing", "running"];
export const RETRIABLE_JOB_STATUSES: JobStatus[] = ["failed", "cancelled", "timeout"];

export interface Job {
  id: string;
  assessment_id: string;
  target_id: string;
  target_value: string;
  tool_id: string;
  tool_name: string;
  status: JobStatus;
  status_message: string | null;
  started_at: string | null;
  completed_at: string | null;
  duration: number | null;
  return_code: number | null;
  retry_count: number;
  log_path: string | null;
  profile_id: string | null;
  generated_command: string[] | null;
  created_at: string;
}

export interface ToolExecutionOptionsPayload {
  profile_id?: string | null;
  advanced_options?: Record<string, unknown> | null;
}

export interface ExecuteRequestPayload {
  tool_names: string[];
  target_ids?: string[] | null;
  tool_options?: Record<string, ToolExecutionOptionsPayload> | null;
}

export interface ServiceResult {
  id: string;
  port: number;
  protocol: string;
  state: string;
  service_name: string | null;
  product: string | null;
  version: string | null;
  extra_info: string | null;
}

export interface HostResult {
  id: string;
  ip_address: string | null;
  hostname: string | null;
  mac_address: string | null;
  mac_vendor: string | null;
  state: string;
  os_name: string | null;
  os_accuracy: number | null;
  services: ServiceResult[];
}

export interface ObservationResult {
  id: string;
  host_id: string | null;
  port: number | null;
  source: string;
  title: string;
  detail: string | null;
}

export interface JobResultsResult {
  job_id: string;
  hosts: HostResult[];
  observations: ObservationResult[];
}

export interface RawOutputResult {
  job_id: string;
  format: string;
  content: string | null;
  created_at: string;
}

export interface ExecuteResult {
  assessment_id: string;
  jobs: Job[];
  queued_count: number;
  skipped_count: number;
}

export interface JobLogsResult {
  job_id: string;
  lines: string[];
  log_path: string | null;
}

export interface AssessmentProgress {
  assessment_id: string;
  total: number;
  pending: number;
  queued: number;
  preparing: number;
  running: number;
  completed: number;
  failed: number;
  cancelled: number;
  timeout: number;
  skipped: number;
  percent_complete: number;
  current_jobs: Job[];
}

export interface JobListParams {
  assessment_id?: string;
  status?: JobStatus;
  tool_name?: string;
  target_id?: string;
  sort_by?: string;
  sort_desc?: boolean;
}

export const JOB_STATUS_OPTIONS: { value: JobStatus; label: string }[] = [
  { value: "pending", label: "Pending" },
  { value: "queued", label: "Queued" },
  { value: "preparing", label: "Preparing" },
  { value: "running", label: "Running" },
  { value: "completed", label: "Completed" },
  { value: "failed", label: "Failed" },
  { value: "cancelled", label: "Cancelled" },
  { value: "timeout", label: "Timeout" },
  { value: "skipped", label: "Skipped" },
];
